File: src/app/vault/fake_client.py
"""Cofre em memória, usado nos testes para não depender do Vault real."""

from __future__ import annotations

import copy
from typing import Any

from app.vault.client import VaultClient
from app.vault.erros import VaultIndisponivelError


class FakeVaultClient(VaultClient):
    def __init__(
        self,
        segredos: dict[str, dict[str, Any]] | None = None,
        *,
        disponivel: bool = True,
    ) -> None:
        self.segredos: dict[str, dict[str, Any]] = copy.deepcopy(segredos or {})
        self.disponivel = disponivel

    def _garantir_disponivel(self) -> None:
        if not self.disponivel:
            raise VaultIndisponivelError("Vault fake configurado como indisponível")

    @staticmethod
    def _normalizar(caminho: str) -> str:
        return caminho.strip("/")

    async def ler_segredo(self, caminho: str) -> dict[str, Any] | None:
        self._garantir_disponivel()
        valor = self.segredos.get(self._normalizar(caminho))
        return copy.deepcopy(valor) if valor is not None else None

    async def gravar_segredo(self, caminho: str, dados: dict[str, Any]) -> None:
        self._garantir_disponivel()
        self.segredos[self._normalizar(caminho)] = copy.deepcopy(dados)

    async def remover_segredo(self, caminho: str) -> None:
        self._garantir_disponivel()
        self.segredos.pop(self._normalizar(caminho), None)

    async def verificar_saude(self) -> bool:
        return self.disponivel

File: src/vault/fake-client.ts
import type { VaultClient } from './client';

/** Cofre em memória — usado em desenvolvimento e nos testes, nunca em produção. */
export class FakeVaultClient implements VaultClient {
  private readonly segredos = new Map<string, Record<string, unknown>>();

  disponivel = true;

  constructor(segredosIniciais?: Record<string, Record<string, unknown>>) {
    for (const [caminho, dados] of Object.entries(segredosIniciais ?? {})) {
      this.segredos.set(this.normalizar(caminho), structuredClone(dados));
    }
  }

  private normalizar(caminho: string): string {
    return caminho.replace(/^\/+|\/+$/g, '');
  }

  private garantirDisponivel(): void {
    if (!this.disponivel) {
      throw new Error('Vault fake configurado como indisponível');
    }
  }

  async lerSegredo(caminho: string): Promise<Record<string, unknown> | null> {
    this.garantirDisponivel();
    const valor = this.segredos.get(this.normalizar(caminho));
    return valor ? structuredClone(valor) : null;
  }

  async gravarSegredo(caminho: string, dados: Record<string, unknown>): Promise<void> {
    this.garantirDisponivel();
    this.segredos.set(this.normalizar(caminho), structuredClone(dados));
  }

  async removerSegredo(caminho: string): Promise<void> {
    this.garantirDisponivel();
    this.segredos.delete(this.normalizar(caminho));
  }

  async verificarSaude(): Promise<boolean> {
    return this.disponivel;
  }

  async existe(caminho: string): Promise<boolean> {
    return (await this.lerSegredo(caminho)) !== null;
  }

  /** Acesso direto ao conteúdo, útil nas asserções dos testes. */
  get segredosArmazenados(): ReadonlyMap<string, Record<string, unknown>> {
    return this.segredos;
  }
}
